presto-accumulo/src/main/java/com/facebook/presto/accumulo/index/ColumnCardinalityCache.java

```diff
@@ -0,0 +1,398 @@
+import java.util.stream.Collectors;
+
+/**
+ * This class is an indexing utility to cache the cardinality of a column value for every table.
+ * Each table has its own cache that is independent of every other, and every column also has its
+ * own Guava cache. Use of this utility can have a significant impact for retrieving the cardinality
+ * of many columns, preventing unnecessary accesses to the metrics table in Accumulo for a
+ * cardinality that won't change much.
+ */
+public class ColumnCardinalityCache
+{
+    private final Authorizations auths;
+    private static final Logger LOG = Logger.get(ColumnCardinalityCache.class);
+    private final Connector conn;
+    private final int size;
+    private final int expireSeconds;
+
+    private Map<String, TableColumnCache> tableToCache = new HashMap<>();
+
+    /**
+     * Creates a new instance of {@link ColumnCardinalityCache}
+     *
+     * @param conn Accumulo connector
+     * @param config Connector configuration for presto
+     * @param auths Authorizations to access Accumulo
+     */
+    public ColumnCardinalityCache(Connector conn, AccumuloConfig config, Authorizations auths)
+    {
+        this.conn = conn;
+        this.size = config.getCardinalityCacheSize();
+        this.expireSeconds = config.getCardinalityCacheExpireSeconds();
+        this.auths = auths;
+    }
+
+    /**
+     * Deletes any cache for the given table, no-op of table does not exist in the cache
+     *
+     * @param schema Schema name
+     * @param table Table name
+     */
+    public void deleteCache(String schema, String table)
+    {
+        LOG.debug("Deleting cache for %s.%s", schema, table);
+        if (tableToCache.containsKey(table)) {
+            // clear the cache and remove it
+            getTableCache(schema, table).clear();
+            tableToCache.remove(table);
+        }
+    }
+
+    /**
+     * Gets the cardinality for each {@link AccumuloColumnConstraint}. Given constraints are
+     * expected to be indexed! Who knows what would happen if they weren't!
+     *
+     * @param schema Schema name
+     * @param table Table name
+     * @param idxConstraintRangePairs Mapping of all ranges for a given constraint
+     * @return A list of
+     * @throws AccumuloException If an error occurs retrieving the cardinalities from Accumulo
+     * @throws AccumuloSecurityException If a security exception is raised
+     * @throws TableNotFoundException If the metrics table does not exist
+     * @throws ExecutionException If another error occurs; I really don't even know anymore.
+     */
+    public List<Pair<AccumuloColumnConstraint, Long>> getCardinalities(String schema, String table,
+            Map<AccumuloColumnConstraint, Collection<Range>> idxConstraintRangePairs)
+            throws AccumuloException, AccumuloSecurityException, TableNotFoundException,
+            ExecutionException
+    {
+        List<Pair<AccumuloColumnConstraint, Long>> cardBuilder = new ArrayList<>();
```

Comment:
I originally had this as an immutable builder, but this list eventually gets sorted by the caller.  Would the preferred practice here be:

1. Leave it as a non-immutable collection
2. Return an immutable collection and have the caller create a mutable copy to be sorted
3. Return the list already sorted and document that it is sorted by cardinality, low to high?